nixos/doc: document fakeNss, binSh (#174093)
NixOS/nixpkgs: doc/builders/images/dockertools.section.md

```

Creating base files like `/etc/passwd` or `/etc/login.defs` is necessary for shadow-utils to manipulate users and groups.

## fakeNss {#ssec-pkgs-dockerTools-fakeNss}

If your primary goal is providing a basic skeleton for user lookups to work,
and/or a lesser privileged user, adding `pkgs.fakeNss` to

> `build*Image.contents` might be the better choice than a custom script running

Reviewer: I'm not convinced that it's better. The supposedly declarative merging of `contents` is a significant source of uncertainty with `dockerTools`. In fact, merging directories and files is non-trivial, so we ought to provide a well-specified and well-documented method for doing so. Shell commands provide that, even if it's hard to admit that `contents` was a failed design.

Author: in the case of `etc/passwd` already being present in another layer, I'd expect the merging (with default priority) to just fail (?).

Shell commands are way less declarative. The code in `examples.nix` before f7ee2706c2aaef359dd05b6910dcbfa6b3493dee, and the `useradd --system --gid nginx nginx` in there wasn't very reproducible - you're not exactly sure what user id `useradd` will pick.

`useradd` and friends.

It provides a `/etc/passwd` and `/etc/group`, containing `root` and `nobody`
users and groups.

It also provides a `/etc/nsswitch.conf`, configuring NSS host resolution to
first check `/etc/hosts`, before checking DNS, as the default in the absence of
a config file (`dns [!UNAVAIL=return] files`) is quite unexpected.

> You usually might to pair it with binSh, which provides `bin/sh` as a symlink
> to `bashInteractive` (as `/bin/sh` is configured as a shell).

Reviewer: I'd expect this to be in a new section for discoverability and focus.

The sentence is a bit broken. You probably meant:

might want to pair

... but I think that's unnecessarily fluffy. How about

```suggestion
You can pair it with `binSh`, which provides `bin/sh` as a symlink
to `bashInteractive` (as `/bin/sh` is configured as a shell).
```

```nix
buildImage {
  name = "shadow-basic";

  contents = [
    binSh
    fakeNss

>   ]

Reviewer: I've deprecated `buildImage` `contents`. You can use `copyToRoot = buildEnv ...` here.
See https://github.com/NixOS/nixpkgs/pull/179801/files
